source/server/admin/prometheus_stats.cc

```diff
@@ -191,59 +192,55 @@ std::string PrometheusStatsFormatter::formattedTags(const std::vector<Stats::Tag
   return absl::StrJoin(buf, ",");
 }
 
-std::string PrometheusStatsFormatter::metricName(const std::string& extracted_name) {
-  std::string sanitized_name = sanitizeName(extracted_name);
-
-  absl::string_view prom_namespace{sanitized_name};
-  prom_namespace = prom_namespace.substr(0, prom_namespace.find_first_of('_'));
-
-  if (prometheusNamespaces().contains(prom_namespace)) {
+absl::optional<std::string>
+PrometheusStatsFormatter::metricName(const std::string& extracted_name,
+                                     const Stats::CustomStatNamespaces& custom_namespaces) {
+  const absl::optional<std::string> custom_namespace_sanitized =
+      custom_namespaces.trySanitizeStatName(extracted_name);
+  if (custom_namespace_sanitized.has_value()) {
+    // This case the name has a custom namespace, and it is a custom metric.
+    const std::string sanitized_name = sanitizeName(custom_namespace_sanitized.value());
+    // We expose these metrics without modifying (e.g. without "envoy_"),
+    // so we have to check the "user-defined" stat name complies with the Prometheus naming
+    // convention. Specifically the name must start with the "[a-zA-Z_]" pattern.
+    // All the characters in sanitized_name are already in "[a-zA-Z0-9_]" pattern
+    // thanks to sanitizeName above, so the only thing we have to do is check
+    // if it does not start with digits.
+    if ('0' <= sanitized_name.front() && sanitized_name.front() <= '9') {
```

Comment:
absl::IsDigit

Also it's worth checking for an empty string first as calling `.front()` on one of those is likely Undefined Behavior.